controller/src/main/java/io/pravega/controller/server/rest/resources/StreamMetadataResourceImpl.java

```diff
@@ -563,11 +564,48 @@ public void listStreams(final String scopeName, final String showInternalStreams
             LoggerHelpers.traceLeave(log, "listStreams", traceId);
             return;
         }
-        boolean showOnlyInternalStreams = showInternalStreams != null && showInternalStreams.equals("true");
-        controllerService.listStreamsInScope(scopeName, requestId)
-                .thenApply(streamsList -> {
-                    StreamsList streams = new StreamsList();
-                    streamsList.forEach((stream, config) -> {
+        boolean showOnlyInternalStreams = filterType != null && filterType.equals("showInternalStreams");
+        boolean showStreamsWithTag = filterType != null && filterType.equals("tag");
+        String tag = null;
+        if (showStreamsWithTag && filterValue != null) {
+            tag = filterValue;
+        }
+        if (tag != null) {
+            List<Stream> streams = new ArrayList<>();
+            StreamsList responseStreams = new StreamsList();
+            String finalTag = tag;
+            localController.listStreamsForTag(scopeName, tag).collectRemaining(s -> streams.add(s)).thenApply(s -> streams).thenApply(s -> {
```

Comment:
please remove the redundant completion stage here.

Reply:
The response object has to be returned in a completable future and cannot be created inside the `forEach` iterator.  The completion stage helps do that.